File: python/queues/queue.py
class Queue:
  def __init__(self, size=10):
    self.size = 10
    self.front, self.rear = -1, -1
    self.queue = [None] * size

  def isEmpty(self):
    return self.front == -1

  def isFull(self):
    return self.size == self.rear+1

  def enqueue(self, e):
    if self.isFull():
      print("Queue Full")
      return 
    elif self.front == -1 and self.rear == -1:
      self.front += 1
      self.rear += 1
      self.queue[self.rear] = e
    else:
      self.rear += 1
      self.queue[self.rear] = e
    print("Added element {} to queue".format(e))

  def peek(self):
    if not self.isEmpty():
      return self.queue[self.front]

  def dequeue(self):
    if self.isEmpty():
      print("Queue is Empty")
    else:
      if self.rear == 0:
        e = self.queue[self.front]
        self.front, self.rear = -1, -1
        return e
      e = self.queue[self.front]
      for i in range(1, self.rear+1):
          self.queue[i-1] = self.queue[i]
      self.rear -= 1
      return e

queue = Queue()
print("Queue Empty: {}".format(queue.isEmpty()))
print("Enqueing element")
for x in range(20):
  if queue.isFull():
    break
  queue.enqueue(x)

print("Queue Full: {}".format(queue.isFull()))
print("Peeking element: {}".format(queue.peek()))
print("Deque elements")
for x in range(20):
  if queue.isEmpty():
    break
  print("Removing Element: {}".format(queue.dequeue()))

File: scala/queues/queue.scala
import scala.util.control.Breaks._

class Queue[T](val size: Int=10){
  private var queue: List[T] = Nil
  private var rear = -1

  def isEmpty = rear == -1

  def isFull = rear+1 == size

  def peek = {
    if(isEmpty) println("Queue is empty") else queue.head 
  }

  def enqueue(e: T) = {
    if(isFull){
      println("Queue is Full")
    }
    else if(isEmpty){
      rear += 1
      queue = queue :+ e
    }
    else{
      rear +=1
      queue = queue :+ e
    }
  }

  def dequeue = {
    if(isEmpty){
      println("Queue is Empty")
    }
    else {
      val ele = queue.head
      queue = queue.tail
      rear -= 1
      ele
    }
  }
}

object QueueTest extends App{
  val queue = new Queue[Int]
  println(s"Queue is empty: ${queue.isEmpty}")
  println("Adding elements to queue")
  for(x <- 1 to queue.size) {
    println(x)
    queue.enqueue(x)
  }
  println(s"Is Queue is Full: ${queue.isFull}")
  println("Removing elemnt from queue")
  println(queue.dequeue)
  println(queue.dequeue)
  println(queue.dequeue)
  println(s"Peeking queue:${queue.peek}")
}
